### PR TITLE
Bugfix/#1039 ExportController teacher authorization vulnerability

### LINKED ISSUE #1039: 🐛 BUG: ExportController teacher authorization vulnerability
The three `ExportController` routes
`{get} /api/export/course/:id`
`{get} /api/export/lecture/:id`
`{get} /api/export/unit/:id`
don't check the authorization of `teacher`s, so any malicious `teacher` can export all courses/lectures/units without hindrance.

## Acceptance criteria:
- [x] Write a fitting access denial unit test for `{get} /api/export/course/:id`.
- [x] Write a fitting access denial unit test for `{get} /api/export/lecture/:id`.
- [x] Write a fitting access denial unit test for `{get} /api/export/unit/:id`.
- [x] Secure `{get} /api/export/course/:id`.
- [x] Secure `{get} /api/export/lecture/:id`.
- [x] Secure `{get} /api/export/unit/:id`.

## Additional info:
The remaining `{get} /api/export/user` route of the `ExportController` seems to be safe: It doesn't take any arbitrary input, instead it directly _(& only)_ exports the `@CurrentUser`.

This security issue was discovered during the recent #853 audits, [see `ExportController`](https://github.com/geli-lms/geli/issues/853#issuecomment-435022945).

### PR DESCRIPTION
**Pull request by torss**
## Description:

Closes #1039

Secures the `ExportController ` by adding the missing `teacher` authorization checks to:
`{get} /api/export/course/:id`
`{get} /api/export/lecture/:id`
`{get} /api/export/unit/:id`

## Core improvements:
- Adds `teacher` authorization checks.
- Adds access denial unit tests _(to ensure that the `teacher` authorization checks work)_.
- Adds 404 checks to the `ExportController` routes.
- Adds `ExportController` 404 unit tests.

## Opportunistic improvements:
- Adds `IUserPrivileges`, `IUserEditPrivileges`, `ICourseUserPrivileges` for the `checkPrivileges`/`checkEditUser`/`checkEditableBy` methods.
- Adds comments to the `duplicate.ts` access denial unit test preparation functions.
- Fixes the generic parameter defaults of `ExtractMongoId`.
- Fixes the static `FixtureUtils` function `getRandomUnitFromCourse`.
- Refactor the export unit tests to use the `TestHelper`.
- Adds 404 checks to the `DuplicationController` routes.
- Adds `DuplicationController` 404 unit tests.
- _Other random minor refactoring._

## Known Issues:
_NONE_

### CODE REVIEW

File: api/src/controllers/DuplicationController.ts [hunk 1]
```diff
@@ -48,13 +48,15 @@ export class DuplicationController {
...
-   * @apiError InternalServerError Failed to duplicate course
+   * @apiError NotFoundError If the course couldn't be found.
+   * @apiError ForbiddenError assertUserDuplicationAuthorization check failed.
    */
   @Post('/course/:id')
   async duplicateCourse(@Param('id') id: string,
                         @BodyParam('courseAdmin', {required: false}) newCourseAdminId: string,
                         @CurrentUser() currentUser: IUser) {
     const courseModel: ICourseModel = await Course.findById(id);
+    if (!courseModel) { throw new NotFoundError(); }
```
[the remark this answers is not in the document]
Comment (torss) [reply api/src/controllers/DuplicationController.ts]: Ok, I had that originally but thought it looked a bit messy, but I will change it back.

[the remark this answers is not in the document]
Comment (danielkesselberg) [reply api/src/controllers/DuplicationController.ts]: You could add a new method for this? Like laravel https://laravel.com/docs/5.7/eloquent#retrieving-single-models did: `findOrFail` they return either the object or throw an exception.

[the remark this answers is not in the document]
Comment (torss) [reply api/src/controllers/DuplicationController.ts]: Good idea! It seems that something already exits in `mongoose`: http://thecodebarbarian.com/whats-new-in-mongoose-53-orfail-and-global-toobject.html :D

[the remark this answers is not in the document]
Comment (torss) [reply api/src/controllers/DuplicationController.ts]: Aw, but it seems we don't have that update in this branch yet?

[the remark this answers is not in the document]
Comment (danielkesselberg) [reply api/src/controllers/DuplicationController.ts]: Sorry :disappointed: Test suite does not pass on mongoose 5.3.x yet.

[the remark this answers is not in the document]
Comment (torss) [reply api/src/controllers/DuplicationController.ts]: Should I change the formatting back to
```js
if (!courseModel) { 
    throw new NotFoundError(); 
}
```
or should we just leave it be and refactor it properly once 5.3.x is compatible?

[the remark this answers is not in the document]
Comment (danielkesselberg) [reply api/src/controllers/DuplicationController.ts]: I leave this up to you. Its only code style.

[the remark this answers is not in the document]
Comment (torss) [reply api/src/controllers/DuplicationController.ts]: Alright, then I will just wait until we can use the neater `.orFail` everywhere.

**Pull request merged**